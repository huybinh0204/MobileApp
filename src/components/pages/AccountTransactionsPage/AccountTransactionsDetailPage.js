import { useTheme } from '@emotion/react';
import { useNavigation } from '@react-navigation/core';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Toast, TransactionItem } from '_components/atoms';
import { AccountBalance, ListItem } from '_components/molecules';
import { SecondaryScreenLayout } from '_components/organisms';
import { TOAST_TYPES } from '_constants';
import strings from '_localization';
import { AccountActions, AccountSelectors } from '_store/account';
import { groupTransactionsByMonth } from '_utilities/Account';
import {
  FullPageSpinner,
  NextPageSpinner,
  SectionHeader,
  SectionList,
} from './AccountTransactionsDetailPage.styles';

const AccountTransactionsDetailPage = () => {
  const { colors } = useTheme();
  const { addListener } = useNavigation();
  const dispatch = useDispatch();

  const accountExternalId = useSelector(AccountSelectors.getExternalId);
  const isLoading = useSelector(AccountSelectors.getIsLoadingTransactions);
  const numberOfPages = useSelector(AccountSelectors.getTransactionsNumberOfPages);
  const transactions = useSelector(AccountSelectors.getTransactions);
  const fetchTransactionsError = useSelector(AccountSelectors.getFetchTransactionsError);
  const [page, setPage] = useState(1);

  const fetchNextPage = () => {
    if (page < numberOfPages) {
      setPage((currentPage) => currentPage + 1);
    }
  };

  const handleCloseToast = () => {
    dispatch(AccountActions.setFetchTransactionsError(null));
  };

  useEffect(() => {
    dispatch(AccountActions.fetchTransactions(accountExternalId, { page, maxCount: 15 }));
  }, [dispatch, accountExternalId, page]);

  useEffect(() => {
    return addListener('blur', () => {
      dispatch(AccountActions.cleanTransactions());
    });
  }, [dispatch, addListener]);

  return (
    <SecondaryScreenLayout title={strings.accounts.allActivity} backgroundColor={colors.background}>
      <AccountBalance />
      {isLoading && transactions.length === 0 ? (
        <FullPageSpinner size="small" color={colors.alpha500} />
      ) : (
        <SectionList
          sections={groupTransactionsByMonth(transactions)}
          keyExtractor={(item, index) => item + index}
          renderItem={({ item }) => <TransactionItem key={item.id} transaction={item} />}
          renderSectionHeader={({ section: { title } }) => <SectionHeader>{title}</SectionHeader>}
          ListFooterComponent={
            isLoading && <NextPageSpinner size="small" color={colors.alpha500} />
          }
          ListEmptyComponent={
            <ListItem color={colors.beta900} title={strings.accounts.emptyState} />
          }
          onEndReached={fetchNextPage}
          onEndReachedThreshold={0.2}
          refreshing={isLoading}
        />
      )}
      <Toast
        header={fetchTransactionsError?.title}
        content={fetchTransactionsError?.description}
        show={fetchTransactionsError !== null}
        onClose={handleCloseToast}
        type={TOAST_TYPES.ERROR}
      />
    </SecondaryScreenLayout>
  );
};

export default AccountTransactionsDetailPage;
