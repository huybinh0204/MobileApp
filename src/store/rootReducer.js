import { combineReducers } from 'redux';
import actionWatchReducer from 'redux-action-watch/lib/reducer';
import { resettableReducer } from 'reduxsauce';
import { AccountReducer } from './account';
import { AuthenticationReducer } from './authentication';
import { CustomerReducer } from './customer';
import { FundDirectDepositStoreReducer } from './pageStore/fundDirectDepositStore';
import { RegisterReducer } from './register';
import { TransferReducer } from './transfer';

const resettable = resettableReducer('RESET');

export default combineReducers({
  account: resettable(AccountReducer),
  auth: resettable(AuthenticationReducer),
  customer: resettable(CustomerReducer),
  fundDirectDepositStoreReducer: resettable(FundDirectDepositStoreReducer),
  register: RegisterReducer,
  transfer: resettable(TransferReducer),
  watcher: actionWatchReducer, // Must be last
});
