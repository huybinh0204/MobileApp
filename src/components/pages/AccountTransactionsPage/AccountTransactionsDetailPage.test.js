import { render } from '@testing-library/react-native';
import React from 'react';
import { createStore } from 'redux';
import rootReducer from '_store/rootReducer';
import ComponentWithProviders from '_test/mocks/ComponentWithProviders';
import AccountTransactionsDetailPage from './AccountTransactionsDetailPage';

describe('AccountTransactionsDetailPage', () => {
  let mockStore;

  beforeEach(() => {
    mockStore = createStore(rootReducer, {
      account: {
        accountInfo: { balance: 1000 },
        transactionsDetails: [],
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const { toJSON } = render(
      <ComponentWithProviders component={AccountTransactionsDetailPage} store={mockStore} />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
