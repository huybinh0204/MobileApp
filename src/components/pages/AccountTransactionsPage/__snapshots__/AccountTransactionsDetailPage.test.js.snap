// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountTransactionsDetailPage should match the snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <RCTSafeAreaView
                        backgroundColor="#FAFAFA"
                        emulateUnlessSupported={true}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "#FAFAFA",
                              "flex": 1,
                            },
                            undefined,
                          ]
                        }
                        testID="SecondaryScreenLayout"
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "height": 108.5,
                                "justifyContent": "center",
                              },
                              undefined,
                            ]
                          }
                        >
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            onClick={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "left": 36,
                                "opacity": 1,
                                "position": "absolute",
                                "zIndex": 10,
                              }
                            }
                            testID="BackButton"
                          >
                            <SvgMock
                              height={45.5}
                              width={45.5}
                            />
                          </View>
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#1F1D25",
                                  "fontFamily": "Poppins-SemiBold",
                                  "fontSize": 29,
                                  "lineHeight": 40,
                                },
                                undefined,
                              ]
                            }
                          >
                            All activity
                          </Text>
                        </View>
                        <View
                          onLayout={[Function]}
                          style={
                            Array [
                              Array [
                                Object {
                                  "flex": 1,
                                },
                                undefined,
                              ],
                              Object {
                                "paddingBottom": 0,
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "justifyContent": "center",
                                  "marginBottom": 29,
                                },
                                undefined,
                              ]
                            }
                          >
                            <Text
                              allowFontScaling={false}
                              isNegative={false}
                              style={
                                Array [
                                  Object {
                                    "color": "#1F1D25",
                                    "fontFamily": "Poppins-SemiBold",
                                    "fontSize": 74.5,
                                    "lineHeight": 103.5,
                                  },
                                  undefined,
                                ]
                              }
                              typography="heading1"
                            >
                              $1,000.00
                            </Text>
                            <Text
                              allowFontScaling={false}
                              style={
                                Array [
                                  Object {
                                    "color": "#5F6A77",
                                    "fontFamily": "Poppins",
                                    "fontSize": 25.5,
                                    "lineHeight": 35.5,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              Available balance
                            </Text>
                          </View>
                          <RCTScrollView
                            ListEmptyComponent={
                              <ListItem
                                color="#1F1D25"
                                icon={null}
                                subtitle={null}
                                title="No activity yet"
                              />
                            }
                            data={Array []}
                            getItem={[Function]}
                            getItemCount={[Function]}
                            keyExtractor={[Function]}
                            onContentSizeChange={[Function]}
                            onEndReached={[Function]}
                            onEndReachedThreshold={0.2}
                            onLayout={[Function]}
                            onMomentumScrollBegin={[Function]}
                            onMomentumScrollEnd={[Function]}
                            onScroll={[Function]}
                            onScrollBeginDrag={[Function]}
                            onScrollEndDrag={[Function]}
                            renderItem={[Function]}
                            scrollEventThrottle={50}
                            stickyHeaderIndices={Array []}
                            style={
                              Array [
                                Object {
                                  "flex": 1,
                                  "paddingHorizontal": 58,
                                },
                                undefined,
                              ]
                            }
                          >
                            <View>
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "center",
                                      "alignSelf": "stretch",
                                      "flexDirection": "row",
                                      "justifyContent": "space-between",
                                    },
                                    undefined,
                                  ]
                                }
                              >
                                <View
                                  style={
                                    Array [
                                      Object {
                                        "alignItems": "center",
                                        "flex": 1,
                                        "flexDirection": "row",
                                      },
                                      undefined,
                                    ]
                                  }
                                >
                                  <View
                                    style={
                                      Array [
                                        Object {
                                          "width": "100%",
                                        },
                                        undefined,
                                      ]
                                    }
                                  >
                                    <Text
                                      allowFontScaling={false}
                                      color="#1F1D25"
                                      numberOfLines={1}
                                      style={
                                        Array [
                                          Object {
                                            "color": "#1F1D25",
                                            "fontFamily": "Poppins-SemiBold",
                                            "fontSize": 25.5,
                                            "lineHeight": 35.5,
                                            "width": "80%",
                                          },
                                          undefined,
                                        ]
                                      }
                                    >
                                      No activity yet
                                    </Text>
                                  </View>
                                </View>
                              </View>
                            </View>
                          </RCTScrollView>
                          <View
                            onLayout={[Function]}
                            paddingHorizontal={null}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "bottom": 0,
                                  "paddingHorizontal": 58,
                                  "position": "absolute",
                                  "width": "100%",
                                },
                                Object {
                                  "opacity": 1,
                                  "transform": Array [
                                    Object {
                                      "translateY": AnimatedValue {
                                        " __value": 0,
                                      },
                                    },
                                  ],
                                },
                              ]
                            }
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "alignSelf": "stretch",
                                    "backgroundColor": "#FFFFFF",
                                    "borderColor": "#F2F3F4",
                                    "borderRadius": 10,
                                    "borderWidth": 0,
                                    "padding": 29,
                                    "shadowColor": "#362E60",
                                    "shadowOffset": Object {
                                      "height": 7,
                                      "width": 0,
                                    },
                                    "shadowOpacity": 0.1,
                                    "shadowRadius": 10,
                                  },
                                  Array [
                                    Object {
                                      "backgroundColor": "#D02B35",
                                      "borderColor": "#D02B35",
                                      "borderWidth": 1,
                                      "width": "100%",
                                    },
                                    undefined,
                                  ],
                                ]
                              }
                              type="error"
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "flex-start",
                                      "flexDirection": "row",
                                      "justifyContent": "space-between",
                                    },
                                    undefined,
                                  ]
                                }
                                testID="toastHeader"
                              >
                                <SvgMock
                                  height={36}
                                  width={36}
                                />
                                <Text
                                  allowFontScaling={false}
                                  numberOfLines={2}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#FFFFFF",
                                        "flex": 1,
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 21.5,
                                        "lineHeight": 30.5,
                                        "paddingHorizontal": 18,
                                        "paddingTop": 2,
                                      },
                                      undefined,
                                    ]
                                  }
                                  type="error"
                                />
                                <View
                                  accessibilityLabel="Close toast"
                                  accessibilityRole="button"
                                  accessible={true}
                                  collapsable={false}
                                  focusable={true}
                                  onClick={[Function]}
                                  onResponderGrant={[Function]}
                                  onResponderMove={[Function]}
                                  onResponderRelease={[Function]}
                                  onResponderTerminate={[Function]}
                                  onResponderTerminationRequest={[Function]}
                                  onStartShouldSetResponder={[Function]}
                                  style={
                                    Object {
                                      "opacity": 1,
                                    }
                                  }
                                  testID="toastCloseButton"
                                >
                                  <SvgMock
                                    height={20}
                                    style={
                                      Array [
                                        Object {
                                          "height": 65,
                                          "width": 65,
                                        },
                                        undefined,
                                      ]
                                    }
                                    type="error"
                                    width={20}
                                  />
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                      </RCTSafeAreaView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
